{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# version 2.0\n",
    "# ugbio_methylation.methyldackel_qc_report.ipynb\n",
    "# =============================================\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# from IPython.display import display, HTML\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from IPython.display import HTML, display  # noqa:A004"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_rows\", 10000)\n",
    "pd.set_option(\"display.width\", 1200)\n",
    "pd.set_option(\"display.max_colwidth\", 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "input_h5_file = \"\"\n",
    "input_base_file_name = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# seperate the processing into the different tables from MethylDackel\n",
    "with pd.HDFStore(input_h5_file, \"r\") as store:\n",
    "    list_tables = store.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "def format_metric_names(df_in):\n",
    "    df_in[\"metric\"] = df_in[\"metric\"].str.replace(r\"PercentMethylation\", \"Percent Methylation: \", regex=True)\n",
    "    df_in[\"metric\"] = df_in[\"metric\"].str.replace(\n",
    "        r\"PercentMethylationPosition\", \"Percent Methylation Position: \", regex=True\n",
    "    )\n",
    "    df_in[\"metric\"] = df_in[\"metric\"].str.replace(r\"CumulativeCoverage\", \"Cumulative Coverage\", regex=True)\n",
    "    df_in[\"metric\"] = df_in[\"metric\"].str.replace(r\"Coverage\", \"Coverage: \", regex=True)\n",
    "    df_in[\"metric\"] = df_in[\"metric\"].str.replace(r\"TotalCpGs\", \"Total CpGs: \", regex=True)\n",
    "    df_in[\"metric\"] = df_in[\"metric\"].str.replace(r\"_\", \" \", regex=True)\n",
    "    return df_in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "def parse_metric_names(df_in):\n",
    "    df_in[\"metric_orig\"] = df_in[\"metric\"]\n",
    "    pat = r\"\\w+_(\\d+)\"  # get value of bins\n",
    "    df_in[\"bin\"] = df_in[\"metric\"].str.extract(pat)\n",
    "    # get metric name\n",
    "    pat = r\"(\\w+)_\\d+\"\n",
    "    df_in[\"metric\"] = df_in[\"metric\"].str.extract(pat)\n",
    "    return df_in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "def display_side_by_side(dfs: list, captions: list, tablespacing=2):\n",
    "    output = \"\"\n",
    "    for caption, df in zip(captions, dfs, strict=False):\n",
    "        output += df.style.set_table_attributes(\"style='display:inline-table'\").set_caption(caption)._repr_html_()\n",
    "        output += tablespacing * \"\\xa0\"\n",
    "\n",
    "    display(HTML(output))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# QC Report for Methylation Calling\n",
    "## Input parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "to_print_sample_info = pd.DataFrame(\n",
    "    data={\"value\": [input_base_file_name, str(input_h5_file)]}, index=[\"Sample name\", \"h5 file\"]\n",
    ")\n",
    "\n",
    "to_print_sample_info[\"value\"] = to_print_sample_info[\"value\"].str.wrap(100)\n",
    "\n",
    "\n",
    "def wrap_df_text(df):\n",
    "    return display(HTML(df.to_html().replace(\"\\\\n\", \"<br>\")))\n",
    "\n",
    "\n",
    "wrap_df_text(to_print_sample_info.style.set_properties(**{\"text-align\": \"left\"}))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Human Genome: Global Methylation Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# main statistics table\n",
    "# --------------------------\n",
    "tbl = \"merge_context_desc\"\n",
    "genome = \"hg\"\n",
    "df_desc = pd.read_hdf(input_h5_file, key=tbl)\n",
    "df_desc = pd.DataFrame(df_desc)\n",
    "df_desc = df_desc.query(\"detail == @genome\")\n",
    "df_desc = df_desc.reset_index()\n",
    "df_desc = df_desc.drop(columns=[\"detail\"])\n",
    "df_desc = format_metric_names(df_desc)\n",
    "to_int = df_desc.metric == \"Total CpGs: \"\n",
    "to_float = df_desc.metric != \"Total CpGs: \"\n",
    "df_desc.loc[to_int, \"value\"] = df_desc.loc[to_int, \"value\"].map(\"{:,.0f}\".format)\n",
    "df_desc.loc[to_float, \"value\"] = df_desc.loc[to_float, \"value\"].map(\"{:,.2f}\".format)\n",
    "display(df_desc.set_index(\"metric\"))\n",
    "\n",
    "# -----------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Human Genome: Details of Per-Read Descriptive Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# per Read information\n",
    "# ------------------------\n",
    "\n",
    "tbl = \"per_read_desc\"\n",
    "genome = \"hg\"\n",
    "df_desc = pd.read_hdf(input_h5_file, key=tbl)\n",
    "df_desc = pd.DataFrame(df_desc)\n",
    "df_desc = df_desc.reset_index()\n",
    "df_desc = df_desc.drop(columns=[\"detail\"])\n",
    "df_desc = format_metric_names(df_desc)\n",
    "df_desc[\"value\"] = df_desc[\"value\"].map(\"{:,.2f}\".format)\n",
    "display(df_desc.set_index(\"metric\"))\n",
    "\n",
    "# -----------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Human Genome: Cytosines in Other Contexts Descriptive Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "tbl = \"merge_context_non_cpg_desc\"\n",
    "\n",
    "df_desc = pd.read_hdf(input_h5_file, key=tbl)\n",
    "df_desc = pd.DataFrame(df_desc)\n",
    "df_desc = df_desc.reset_index()\n",
    "df_desc = format_metric_names(df_desc)\n",
    "\n",
    "df_desc[\"stat_type\"] = df_desc[\"metric\"].str.extract(r\"([A-Za-z]+)[\\s:]\")\n",
    "df_desc[\"metric\"] = df_desc[\"metric\"].str.title()\n",
    "df_desc[\"value\"][df_desc[\"stat_type\"] == \"Percent\"] = (df_desc[\"value\"][df_desc[\"stat_type\"] == \"Percent\"] / 100).map(\n",
    "    \"{:,.2%}\".format\n",
    ")\n",
    "df_desc[\"value\"][df_desc[\"stat_type\"] == \"Coverage\"] = df_desc[\"value\"][df_desc[\"stat_type\"] == \"Coverage\"].map(\n",
    "    \"{:,.2f}\".format\n",
    ")\n",
    "\n",
    "table_names = df_desc[\"detail\"].unique()\n",
    "df_output = []\n",
    "df_output = [y for x, y in df_desc.groupby(\"detail\")]\n",
    "\n",
    "df_to_print = []\n",
    "for line in df_output:\n",
    "    line.index = line[\"metric\"]\n",
    "    line = line[\"value\"].to_frame()  # noqa: PLW2901\n",
    "    df_to_print.append(line)\n",
    "\n",
    "display_side_by_side(df_to_print, table_names)\n",
    "\n",
    "# --------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# function for creating Mbias plots\n",
    "# -----------------------------------------------------------------------------------\n",
    "def plot_mbias(in_list_df):\n",
    "    plt.style.use(\"ggplot\")\n",
    "\n",
    "    if len(in_list_df) == 4:  # noqa: PLR2004\n",
    "        i = j = k = 0\n",
    "        in_colours = [\"tomato\", \"indianred\", \"tomato\", \"indianred\"]\n",
    "        f, ax = plt.subplots(2, 2, figsize=[12, 12])\n",
    "\n",
    "        for j in range(ax.shape[0]):\n",
    "            for k in range(ax.shape[1]):\n",
    "                currax = ax[j][k]\n",
    "                curr_title = in_list_df[i][\"detail\"].unique()[0]\n",
    "\n",
    "                sns.lineplot(data=in_list_df[i], x=\"bin\", y=\"value\", lw=2.5, ax=currax, color=in_colours[i])\n",
    "                currax.set_xlabel(\"Position\", fontsize=14)\n",
    "                currax.set_ylabel(\"Fraction of Methylation\", fontsize=14)\n",
    "                currax.set_title(curr_title, fontsize=14)\n",
    "                currax.tick_params(labelsize=14)\n",
    "                plt.xticks(rotation=45)\n",
    "                currax.set_ylim([0, 1])\n",
    "                i += 1\n",
    "\n",
    "        plt.tight_layout()\n",
    "\n",
    "    else:\n",
    "        in_colours = [\"tomato\", \"indianred\"]\n",
    "        f, ax = plt.subplots(1, 2, figsize=[12, 5.5])\n",
    "\n",
    "        for i in range(len(in_list_df)):\n",
    "            currax = ax[i]\n",
    "            curr_title = in_list_df[i][\"detail\"].unique()[0]\n",
    "\n",
    "            sns.lineplot(data=in_list_df[i], x=\"bin\", y=\"value\", lw=2.5, ax=currax, color=in_colours[i])\n",
    "            currax.set_xlabel(\"Position\", fontsize=14)\n",
    "            currax.set_ylabel(\"Fraction of Methylation\", fontsize=14)\n",
    "            currax.set_title(list_tables[i], fontsize=14)\n",
    "            currax.tick_params(labelsize=14)\n",
    "            plt.xticks(rotation=45)\n",
    "            currax.set_ylim([0, 1])\n",
    "\n",
    "        plt.tight_layout()\n",
    "\n",
    "\n",
    "# --------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Human Genome: M-bias plots of mean methylation along reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# Mbias: Mean methylation along reads\n",
    "# ========================================\n",
    "\n",
    "tbl = \"mbias_per_position\"\n",
    "\n",
    "df_pos = pd.read_hdf(input_h5_file, key=tbl)\n",
    "df_pos = pd.DataFrame(df_pos)\n",
    "df_pos = df_pos.reset_index()\n",
    "df_pos = parse_metric_names(df_pos)\n",
    "df_pos[\"bin\"] = df_pos[\"bin\"].astype(int)\n",
    "list_tables = list(set(df_pos[\"detail\"]))\n",
    "\n",
    "df_to_print = df_pos.copy()\n",
    "df_to_print[\"stat_type\"] = df_to_print[\"metric\"].str.extract(r\"([A-Za-z]+)\\s\")\n",
    "df_to_print[\"metric\"] = df_to_print[\"metric\"].str.title()\n",
    "\n",
    "list_df = [y for x, y in df_to_print.groupby(df_pos[\"detail\"], sort=False)]\n",
    "\n",
    "# plot the MBIAS tests\n",
    "plot_mbias(list_df)\n",
    "\n",
    "# ---------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## M-bias Descriptive Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# CONTINUES Mbias: Mean methylation along reads\n",
    "tbl = \"mbias_desc\"\n",
    "\n",
    "df_desc = pd.read_hdf(input_h5_file, key=tbl)\n",
    "df_desc = pd.DataFrame(df_desc)\n",
    "df_desc = df_desc.reset_index()\n",
    "df_desc = format_metric_names(df_desc)\n",
    "\n",
    "df_to_print = df_desc.copy()\n",
    "df_to_print[\"stat_type\"] = df_to_print[\"metric\"].str.extract(r\"([A-Za-z]+)\\s\")\n",
    "df_to_print[\"metric\"] = df_to_print[\"metric\"].str.title()\n",
    "\n",
    "\n",
    "df_to_print[\"value\"][df_to_print[\"stat_type\"] == \"Percent\"] = (\n",
    "    df_to_print[\"value\"][df_to_print[\"stat_type\"] == \"Percent\"]\n",
    ").map(\"{:,.2%}\".format)\n",
    "\n",
    "\n",
    "cols = [\"metric\", \"value\"]\n",
    "df_output = []\n",
    "df_output = [y for x, y in df_to_print.groupby(\"detail\")]\n",
    "\n",
    "\n",
    "if len(df_output) == 4:  # noqa: PLR2004\n",
    "    order = [3, 2, 1, 0]\n",
    "else:\n",
    "    order = [1, 0]\n",
    "df_output = [df_output[i] for i in order]\n",
    "\n",
    "del df_to_print\n",
    "df_to_print = []\n",
    "temp_tables = []\n",
    "for line in df_output:\n",
    "    line.index = line[\"metric\"]\n",
    "    temp_tables.append(line[\"detail\"][0])\n",
    "    df_to_print.append(line[\"value\"].to_frame())\n",
    "\n",
    "display_side_by_side(df_to_print, temp_tables)\n",
    "\n",
    "# ---------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Human Genome: M-bias plots of mean methylation on CHH/CHG along reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MbiasNoCpG: Mean methylation along reads on non-CpG Cytosines\n",
    "# ========================================\n",
    "\n",
    "tbl = \"mbias_non_cpg_per_position\"\n",
    "\n",
    "df_pos = pd.read_hdf(input_h5_file, key=tbl)\n",
    "df_pos = pd.DataFrame(df_pos)\n",
    "df_pos = df_pos.reset_index()\n",
    "df_pos = parse_metric_names(df_pos)\n",
    "df_pos[\"bin\"] = df_pos[\"bin\"].astype(int)\n",
    "list_tables = list(set(df_pos[\"detail\"]))\n",
    "\n",
    "df_to_print = df_pos.copy()\n",
    "df_to_print[\"stat_type\"] = df_to_print[\"metric\"].str.extract(r\"([A-Za-z]+)\\s\")\n",
    "df_to_print[\"metric\"] = df_to_print[\"metric\"].str.title()\n",
    "\n",
    "list_df = [y for x, y in df_to_print.groupby(df_pos[\"detail\"], sort=False)]\n",
    "\n",
    "# plot the MBIAS tests\n",
    "plot_mbias(list_df)\n",
    "\n",
    "# ---------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## M-bias of CHH/CHG Descriptive Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CONTINUES MbiasNoCpG: Mean methylation along reads on non-CpG Cytosines\n",
    "tbl = \"mbias_non_cpg_desc\"\n",
    "\n",
    "df_desc = pd.read_hdf(input_h5_file, key=tbl)\n",
    "df_desc = pd.DataFrame(df_desc)\n",
    "df_desc = df_desc.reset_index()\n",
    "df_desc = format_metric_names(df_desc)\n",
    "\n",
    "df_to_print = df_desc.copy()\n",
    "df_to_print[\"stat_type\"] = df_to_print[\"metric\"].str.extract(r\"([A-Za-z]+)\\s\")\n",
    "df_to_print[\"metric\"] = df_to_print[\"metric\"].str.title()\n",
    "\n",
    "\n",
    "df_to_print[\"value\"][df_to_print[\"stat_type\"] == \"Percent\"] = (\n",
    "    df_to_print[\"value\"][df_to_print[\"stat_type\"] == \"Percent\"]\n",
    ").map(\"{:,.2%}\".format)\n",
    "\n",
    "\n",
    "cols = [\"metric\", \"value\"]\n",
    "df_output = []\n",
    "df_output = [y for x, y in df_to_print.groupby(\"detail\")]\n",
    "\n",
    "\n",
    "if len(df_output) == 4:  # noqa: PLR2004\n",
    "    order = [3, 2, 1, 0]\n",
    "else:\n",
    "    order = [1, 0]\n",
    "df_output = [df_output[i] for i in order]\n",
    "\n",
    "del df_to_print\n",
    "df_to_print = []\n",
    "temp_tables = []\n",
    "for line in df_output:\n",
    "    line.index = line[\"metric\"]\n",
    "    temp_tables.append(line[\"detail\"][0])\n",
    "    df_to_print.append(line[\"value\"].to_frame())\n",
    "\n",
    "display_side_by_side(df_to_print, temp_tables)\n",
    "\n",
    "# ---------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "## Control genomes (if exist)\n",
    "# ------------------------------\n",
    "\n",
    "tbl = \"merge_context_per_position\"\n",
    "\n",
    "df_pos = pd.read_hdf(input_h5_file, key=tbl)\n",
    "df_pos = pd.DataFrame(df_pos)\n",
    "df_pos = df_pos.reset_index()\n",
    "all_genomes = list(set(df_pos[\"detail\"]))\n",
    "ctrl_genomes = [\"Lambda\", \"pUC19\"]\n",
    "\n",
    "check = all(item in all_genomes for item in ctrl_genomes)\n",
    "\n",
    "if check is True:\n",
    "    display(HTML(\"<h2>\" + \"Control Genomes: Methylation and Coverage\" + \"</h2>\"))\n",
    "\n",
    "    # PRINT PLOTS OF PERCENT METHYLATION ACROSS ENTIRE CONTROL GENOMES\n",
    "    # -----------------------------------------------------------------\n",
    "\n",
    "    df_pos = parse_metric_names(df_pos)\n",
    "    df_pos[\"bin\"] = df_pos[\"bin\"].astype(int)\n",
    "    df_pos = format_metric_names(df_pos)\n",
    "    df_pos[\"bin\"] = df_pos[\"bin\"].astype(int)\n",
    "    df_pos = df_pos.query('metric == \"Percent Methylation: Position\"')\n",
    "    df_output = [y for x, y in df_pos.groupby(\"detail\")]\n",
    "\n",
    "    df_pos_meth = []\n",
    "    n = 102\n",
    "    f, ax = plt.subplots(1, 2, figsize=[12, 5])\n",
    "    i = 0\n",
    "    palet = [\"forestgreen\", \"steelblue\"]\n",
    "\n",
    "    for df_pos_meth in df_output:\n",
    "        # get methylation per position\n",
    "        df_pos_meth = df_pos_meth.reset_index()  # noqa: PLW2901\n",
    "        temp_genome = df_pos_meth[\"detail\"].unique()[0]\n",
    "\n",
    "        # print to subplots\n",
    "        currax = ax[i]\n",
    "        s = df_pos_meth.plot(\n",
    "            kind=\"area\",\n",
    "            ylim=[0, n],\n",
    "            y=\"value\",\n",
    "            title=temp_genome + \": Percent methylation\",\n",
    "            legend=False,\n",
    "            color=palet[i],\n",
    "            alpha=0.6,\n",
    "            ax=currax,\n",
    "            fontsize=14,\n",
    "        )\n",
    "        a = currax.set(xlabel=\"Position\", ylabel=\"Percent Methylation\")\n",
    "        plt.style.use(\"ggplot\")\n",
    "\n",
    "        i += 1\n",
    "\n",
    "# --------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "## Control genomes (if exist)\n",
    "# ------------------------------\n",
    "\n",
    "tbl = \"merge_context_desc\"\n",
    "df_desc = pd.read_hdf(input_h5_file, key=tbl)\n",
    "df_desc = pd.DataFrame(df_desc)\n",
    "df_desc = df_desc.reset_index()\n",
    "all_genomes = list(set(df_desc[\"detail\"]))\n",
    "ctrl_genomes = [\"Lambda\", \"pUC19\"]\n",
    "\n",
    "check = all(item in all_genomes for item in ctrl_genomes)\n",
    "\n",
    "if check is True:\n",
    "    display(HTML(\"<h2>\" + \"Control Genomes: Methylation and Coverage Descriptive Statistics\" + \"</h2>\"))\n",
    "\n",
    "    tbl = \"merge_context\"\n",
    "    genome = \"hg\"\n",
    "    df_to_print = pd.DataFrame()\n",
    "    df_table = df_desc[df_desc[\"detail\"] != genome]\n",
    "    df_table = df_table.reset_index()\n",
    "    df_table = format_metric_names(df_table)\n",
    "\n",
    "    df_table[\"stat_type\"] = df_table[\"metric\"].str.extract(r\"([A-Za-z]+)[\\s:]\")\n",
    "    df_table[\"metric\"] = df_table[\"metric\"].str.title()\n",
    "    df_table[\"value\"][df_table[\"stat_type\"] == \"Percent\"] = (\n",
    "        df_table[\"value\"][df_table[\"stat_type\"] == \"Percent\"] / 100\n",
    "    ).map(\"{:,.2%}\".format)\n",
    "    df_table[\"value\"][df_table[\"stat_type\"] == \"Coverage\"] = df_table[\"value\"][df_table[\"stat_type\"] == \"Coverage\"].map(\n",
    "        \"{:,.2f}\".format\n",
    "    )\n",
    "    df_table[\"value\"][df_table[\"stat_type\"] == \"Total\"] = df_table[\"value\"][df_table[\"stat_type\"] == \"Total\"].map(\n",
    "        \"{:,.0f}\".format\n",
    "    )\n",
    "    df_table[\"metric\"] = df_table[\"metric\"].str.replace(r\"Cpgs\", \"CpGs\", regex=True)\n",
    "\n",
    "    table_names = df_table[\"detail\"].unique()\n",
    "    cols = [\"metric\", \"value\", \"detail\"]\n",
    "    df_output = []\n",
    "    df_output = [y for x, y in df_table.groupby(\"detail\")]\n",
    "\n",
    "    df_to_print = []\n",
    "    for line in df_output:\n",
    "        line.index = line[\"metric\"]\n",
    "        df_to_print.append(line[\"value\"].to_frame())\n",
    "\n",
    "    display_side_by_side(df_to_print, table_names)\n",
    "\n",
    "    # --------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# function for printing bar plots of methylation and coverage at CpGs\n",
    "def plot_bar_distrib(in_table):\n",
    "    i = 0\n",
    "    in_colours = [\"salmon\", \"tomato\"]\n",
    "    curr_genome = \"Human Genome\"\n",
    "    n_rows = in_table.shape[0]\n",
    "    if n_rows > 10:  # noqa: PLR2004\n",
    "        h = 5\n",
    "        w = 13.5\n",
    "        n = 2\n",
    "\n",
    "        in_list_df = [y for x, y in in_table.groupby(in_table[\"metric\"], sort=False)]\n",
    "        f, ax = plt.subplots(1, n, figsize=[w, h])  # noqa: F821\n",
    "        # noqa: F821\n",
    "        for i in range(len(in_list_df)):\n",
    "            currax = ax[i]\n",
    "            y_axis_label = in_list_df[i][\"metric\"].unique()[0]\n",
    "            y_axis_label = y_axis_label.replace(\":\", \"\")\n",
    "            x_axis_label = \"Value Bins\"\n",
    "            curr_title = curr_genome + \": \" + y_axis_label\n",
    "\n",
    "            sns.barplot(data=in_list_df[i], x=\"bin\", y=\"value\", lw=2.5, ax=currax, color=in_colours[i])\n",
    "            currax.set_xlabel(x_axis_label, fontsize=14)\n",
    "            currax.set_ylabel(y_axis_label, fontsize=14)\n",
    "            currax.set_title(curr_title, fontsize=13)\n",
    "            currax.tick_params(labelsize=14)\n",
    "            f.axes[i].tick_params(labelrotation=45)\n",
    "\n",
    "    else:\n",
    "        h = 5\n",
    "        w = 5.5\n",
    "        n = 1\n",
    "\n",
    "        f, ax = plt.subplots(n, n, figsize=[w, h])\n",
    "\n",
    "        y_axis_label = in_table[\"metric\"].unique()[0]\n",
    "        y_axis_label = y_axis_label.replace(\":\", \"\")\n",
    "        curr_title = curr_genome + \": \" + y_axis_label\n",
    "\n",
    "        sns.barplot(data=in_table, x=\"bin\", y=\"value\", lw=2.5, ax=ax, color=in_colours[i])\n",
    "        ax.set_xlabel(\"Value Bins\", fontsize=14)\n",
    "        ax.set_ylabel(y_axis_label, fontsize=14)\n",
    "        ax.set_title(curr_title, fontsize=13)\n",
    "        ax.tick_params(labelsize=14)\n",
    "        plt.xticks(rotation=45)\n",
    "\n",
    "\n",
    "# --------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# print details of mergeCOntext plots of CpG Methylation and Coverage\n",
    "# -----------------------------\n",
    "# main statistics table\n",
    "# --------------------------\n",
    "tbl = \"merge_context_hist\"\n",
    "genome = \"hg\"\n",
    "df_hist = pd.read_hdf(input_h5_file, key=tbl)\n",
    "df_hist = pd.DataFrame(df_hist)\n",
    "df_hist = df_hist.reset_index()\n",
    "df_hist = df_hist[df_hist[\"detail\"] == genome]\n",
    "df_hist = parse_metric_names(df_hist)\n",
    "df_hist = format_metric_names(df_hist)\n",
    "df_hist[\"stat_type\"] = df_hist[\"metric\"].str.extract(r\"([A-Za-z]+):\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Human Genome: Additional Details of CpG Methylation and Coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_df = [y for x, y in df_hist.groupby(df_hist[\"stat_type\"], sort=False)]\n",
    "\n",
    "for df_to_plot in list_df:\n",
    "    plot_bar_distrib(df_to_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Human Genome: Values of CpG Methylation and Coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print details of figures\n",
    "# --------------------------\n",
    "\n",
    "list_df = [y for x, y in df_hist.groupby(df_hist[\"metric\"], sort=False)]\n",
    "\n",
    "df_to_print = []\n",
    "temp_tables = []\n",
    "for i, sub_df in enumerate(list_df):\n",
    "    sub_df = sub_df.reset_index()  # noqa: PLW2901\n",
    "    if i < 2:  # noqa: PLR2004\n",
    "        sub_df[\"value\"] = sub_df[\"value\"].map(\"{:,.0f}\".format)\n",
    "    else:\n",
    "        sub_df[\"value\"] = sub_df[\"value\"].map(\"{:,.2%}\".format)\n",
    "    temp_tables.append(sub_df[\"metric\"][0].replace(\":\", \"\"))\n",
    "    sub_df.index = sub_df[\"bin\"]\n",
    "    df_to_print.append(sub_df[\"value\"].to_frame())\n",
    "\n",
    "\n",
    "display_side_by_side(df_to_print, temp_tables)\n",
    "# -------"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
